Src/IronPython/Runtime/Operations/StringOps.cs

```diff
@@ -1577,19 +1555,23 @@ public static class StringOps {
             // in the common case we don't need to encode anything, so we
             // lazily create the StringBuilder only if necessary.
             StringBuilder b = null;
-            for (int i = 0; i < s.Length; i++) {
-                char ch = s[i];
-                if (ch <= 0x1f || ch == 0x7f) {
-                    ReprInit(ref b, s, i);
+            int start = 0;
+            int count = s.Length;
+            int i = start;
+            while (i < s.Length) {
+                char ch = s[i++];
+                if (ch <= 0x1F || ch == 0x7F) {
+                    StringBuilderInit(ref b, s, start, i - 1);
```

Comment:
Good observation, it caught my eye too to the point that I delayed this submission to think more about it. Here is the full story: 

The previous implementation of `StringBuilderInit` (called `ReprInit`) did not allow for non-zero start point so it needed to be upgraded. Originally I simply followed the `StringBuilder` initialization pattern from `LiteralParser`, which has parameters: string to use for initialization, start index of the substrring, and count (i.e. number of characters in the substring) (plus initial capacity, but this aside for now). 

But this leads to **every** call to `StringBuilderInit` having form

```C#
StringBuilderInit(ref b, s, start, i - start - 1);
```

Which is a bummer considering that every call to `StringBuilderInit` except the very first one is a no-op and return, but on every call the whole argument expression `i - start - 1` has to be evaluated. A simple solution to this is to push the whole evaluation down into `StringBuilderInit`, after the null test, but it would lead to a very bizarre and error-prone signature of the method. So instead of that I settled on changing  `count` to exclusive `end`, which is a common pattern. This reduces `i - start - 1` to `i - 1`. Performance wise, it is much better as `-1` can be done by a single decrement instruction, rather than fetching another variable and performing the calculation.

The next step was to look how it would work if `i` were counted so that it already had the right value for the call. As it is now, the invariant is that `i` is the index of the next unprocessed character and advances as soon as the character has been read. It is possible to change it so that it becomes the index of the character being analyzed. But then the peek at the next character will be `i + 1` here and there. I tried that form but I found the code to be less readable. Also I believe it is good to use the same indexing principle (whichever is chosen) for the whole group of methods that do encoding/decoding of escapes (which includes `AsciiEncode`, `RawUnicodeEscapeEncode` in `StringOps`, `DoParseString`, `ParseBytes` in `LiteralParser`, and `escape_encode`, `escape_decode` in `_codecs`). In the end I decided it was not worth changing it.

Note that every call to `StringBuilderInit` in `LiteralParser.DoParseString` also has the same `i - start - 1`. It does not stand out because the two calls disappear in the code sprawl, but for the sake of consistency and to avoid mistakes I intend to align the signatures of `LiteralParser.StringBuilderInit` and `StringOps.StringBuilderInit` in a future commit.